subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -0,0 +1,299 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+/**
+ * SECTION:tracer-framerate
+ * @short_description: log processing framerate stats
+ *
+ * A tracing module that determines the pipeline framerate and buffer
+ * drop rate. This elements supports tracing the entire pipeline framerate.
+ *
+ * ```
+ * GST_TRACERS="framerate" GST_DEBUG=GST_TRACER:7 ./...
+ * ```
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "gstframerate.h"
+
+GST_DEBUG_CATEGORY_STATIC (gst_framerate_debug);
+#define GST_CAT_DEFAULT gst_framerate_debug
+
+#define _do_init \
+    GST_DEBUG_CATEGORY_INIT (gst_framerate_debug, "framerate", 0, "framerate tracer");
+#define gst_framerate_tracer_parent_class parent_class
+G_DEFINE_TYPE_WITH_CODE (GstFramerateTracer, gst_framerate_tracer,
+    GST_TYPE_TRACER, _do_init);
+
+#define DEFAULT_FPS_UPDATE_INTERVAL_MS 500      /* 500 ms */
+
+static GQuark framerate_probe_id;
+static GQuark sub_framerate_probe_id;
+static GQuark drop_sub_framerate_quark;
+
+static GstTracerRecord *tr_framerate;
+static GstTracerRecord *tr_element_framerate;
+static GstTracerRecord *tr_element_reported_framerate;
+
+static GstPad *
+get_peer_pad (GstPad *pad)
+{
+  GstPad *peer_pad;
+  GstPad *tmp_pad;
+
+  peer_pad = gst_pad_get_peer (pad);
+
+  while (GST_IS_GHOST_PAD (peer_pad)) {
+    tmp_pad = gst_ghost_pad_get_target (peer_pad);
+    gst_object_unref (peer_pad);
+    peer_pad = tmp_pad;
+  }
+
+  return peer_pad;
+}
+
+/* hooks */
+
+
+static gboolean
+gst_framerate_tracer_display_current_fps (GstFramerateTracer *self,
+    GstClockTime current_ts)
+{
+  guint64 frames_rendered, frames_dropped;
+  gdouble rr, dr, average_fps;
+  gdouble time_diff, time_elapsed;
+
+  frames_rendered = g_atomic_int_get (&self->frames_rendered);
+  frames_dropped = g_atomic_int_get (&self->frames_dropped);
+
+  if ((frames_rendered + frames_dropped) == 0) {
+    /* in case timer fired and we didn't yet get any QOS events */
+    return TRUE;
+  }
+
+  time_diff = (gdouble) (current_ts - self->last_ts) / GST_SECOND;
+  time_elapsed = (gdouble) (current_ts - self->start_ts) / GST_SECOND;
+
+  rr = (gdouble) (frames_rendered - self->last_frames_rendered) / time_diff;
+  dr = (gdouble) (frames_dropped - self->last_frames_dropped) / time_diff;
+
+  average_fps = (gdouble) frames_rendered / time_elapsed;
+
+  if (rr > self->max_fps) {
+    self->max_fps = rr;
+    GST_DEBUG_OBJECT (self, "Updated max-fps to %f", rr);
+  }
+  if (rr < self->min_fps) {
```

Comment:
really sorry that I requested that change, but in this line we actually need to either bring back `self->min_fps == -1 ||` or init `self->min_fps` with some big number such as `G_MAXDOUBLE`, so it will take the `rr` on the first run.

Reply:
brought it back